sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/EventProcessor{TPartition}.cs

```diff
@@ -406,6 +410,250 @@ protected EventProcessor()
+        ///
+        /// <param name="partition">The Event Hub partition whose processing should be started.</param>
+        /// <param name="eventBatch">The batch of events to process.</param>
+        /// <param name="dispatchEmptyBatches"><c>true</c> if empty batches should be dispatched to the handler; otherwise, <c>false</c>.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken"/> instance to signal the request to cancel the processing.</param>
+        ///
+        internal virtual async Task ProcessEventBatchAsync(TPartition partition,
+                                                           IEnumerable<EventData> eventBatch,
+                                                           bool dispatchEmptyBatches,
+                                                           CancellationToken cancellationToken)
+        {
+            cancellationToken.ThrowIfCancellationRequested<TaskCanceledException>();
+            eventBatch = (eventBatch as IList<EventData>) ?? eventBatch.ToList();
+
+            // If there were no events in the batch and empty batches should not be emitted,
+            // take no further action.
+
+            if ((!eventBatch.Any()) && (!dispatchEmptyBatches))
+            {
+                return;
+            }
+
+            // Create the diagnostics scope used for distributed tracing and instrument the events in the batch.
+
+            using var diagnosticScope = EventDataInstrumentation.ScopeFactory.CreateScope(DiagnosticProperty.EventProcessorProcessingActivityName);
+            diagnosticScope.AddAttribute(DiagnosticProperty.KindAttribute, DiagnosticProperty.ConsumerKind);
+            diagnosticScope.AddAttribute(DiagnosticProperty.EventHubAttribute, EventHubName);
+            diagnosticScope.AddAttribute(DiagnosticProperty.EndpointAttribute, FullyQualifiedNamespace);
+
+            string diagnosticId;
+
+            if ((diagnosticScope.IsEnabled) && (eventBatch.Any()))
+            {
+                foreach (var eventData in eventBatch)
+                {
+                    if (EventDataInstrumentation.TryExtractDiagnosticId(eventData, out diagnosticId))
+                    {
+                        var attributes = new Dictionary<string, string>(1)
+                        {
+                            { DiagnosticProperty.EnqueuedTimeAttribute, eventData.EnqueuedTime.ToUnixTimeMilliseconds().ToString() }
+                        };
+
+                        diagnosticScope.AddLink(diagnosticId, attributes);
+                    }
+                }
+            }
+
+            diagnosticScope.Start();
+
+            // Dispatch the batch to the handler for processing.  Exceptions in the handler code are intended to be
+            // unhandled by the processor; after logging, signal the exception using the
+
+            try
+            {
+                await OnProcessingEventBatchAsync(eventBatch, partition, cancellationToken).ConfigureAwait(false);
```

Comment:
Sorry for the open ended question here - but do you think this does the right thing with respect to cancellation?  If the cancellation token becomes signaled and the user's code does something like `ThrowIfCancellationRequested`, are we going to the right thing here?  I figure that we'll wrap the exception in a DeveloperCodeException and then later unwrap it and rethrow.  Will it go unhandled at that point?  I know in the past we said if the user's processing code throws an exception we'll just let nature take it's course, but I'm wondering if that's the right thing in light of cancellation here.

I don't actually know what the right answer is - and I'm partially asking about this just to see if you had thought about this.

Reply:
If I'm being honest, I wasn't a fan of including a cancellation token for the handlers for that reason.   Since the caller owns the code managing the processor as well as the handlers, I would have assumed there would be visibility for the intent to cancel.   That said, I really don't know what is more "right" in this case.   Currently, the flow is as you describe.   To restate that, what happens now is:

- We treat the handler as opaque and do not treat cancellation from within it specially; this unifies our flow and makes it explicit in the logs that the decision was made in handler code.  The end result is the same and the same exception type _(TaskCanceledException)_ gets propagated

The alternatives  that I see are:  

- We pass `CancellationToken.None` and make the decision to never explicitly signal cancellation for a batch.   _(I don't think that would be my vision of the right thing)_

- We don't wrap cancellation if the token was signaled and bubble it in its natural state, which would potentially obscure that the handler was the path that made the decision to honor the token, as opposed to cancellation being honored elsewhere.

After writing that, I still don't have a strong feeling, but I think that I'm in favor of our current path over the alternatives.   Thoughts?

Reply:
> We pass CancellationToken.None and make the decision to never explicitly signal cancellation for a batch. (I don't think that would be my vision of the right thing)

Agree on this.

>  which would potentially obscure that the handler was the path that made the decision to honor the token, as opposed to cancellation being honored elsewhere.

Would obscure things?  Shouldn't the stack trace for the exception that was thrown would include the handler (since it originated the exception) and that would be enough to understand what observed the cancellation and raised the exception?

I think the main difference between the two approaches is around if `OnProcessingErrorAsync` is called or not (I think today we do call it, and in the third model we would not).

I still don't have a strong feeling either way - it's clear you've thought about this, so I'm happy with whatever you end up picking.

Reply:
> Would obscure things? Shouldn't the stack trace for the exception that was thrown would include the handler (since it originated the exception) and that would be enough to understand what observed the cancellation and raised the exception?

Truth, so long as someone is inspecting the exception.   I was thinking more of logging, where we're capturing the error messages, but not full stack trace _(as was the track one convention and the one that I've seen most often for ETW)_.   We're using a specific format mask to flag developer-code exceptions, which is what I was thinking about losing when I said "obscuring."

Since it seems we're good with either approach, I'm going to stick with the current to optimize for the log.   Happy to revisit in the future if we wanted to think on expanding our logging story and potentially make tweaks.